File: code/emf/src/cutAndTotal/Cut.java
package cutAndTotal;

import company.*;

/**
 * Cut all salaries by half.
 * 
 */
public class Cut {

	public static void cutCompany(Company company) {
		for (Dept dept : company.getDepts())
			cutDept(dept);
	}

	private static void cutDept(Dept dept) {
		cutEmployee(dept.getManager());
		for (Subunit subunit : dept.getSubunits())
			cutSubunit(subunit);
	}

	private static void cutSubunit(Subunit subunit) {
		if (subunit.getPu() != null)
			cutEmployee(subunit.getPu());
		else
			cutDept(subunit.getDu());
	}

	private static void cutEmployee(Employee employee) {
		employee.setSalary(employee.getSalary() / 2);
	}

}

File: code/haskell/alpha/Cut.hs
module Cut where

import Company

cut :: Company -> Company

cut (n,ds) = (n,map cutDept ds)
 where
  cutDept :: Dept -> Dept
  cutDept (Dept n m sus) = Dept n (cutEmp m) (map cutSu sus)
  cutEmp :: Employee -> Employee
  cutEmp (Employee n a s) = Employee n a (s/2)  
  cutSu :: SubUnit -> SubUnit
  cutSu (PU e) = PU (cutEmp e)
  cutSu (DU d) = DU (cutDept d)
